specifications/Prisoners_Single_Switch/Prisoner.tla

```diff
@@ -0,0 +1,178 @@
+
+  Light_Unknown
+  \* Configuration if the initial state of the light is "off" or
+  \* unknown. The puzzle is harder (takes more steps) if we don't know
+  \* the initial state.
+
+ASSUME Light_Unknown \in BOOLEAN
+ASSUME Designated_Counter \in Prisoner
+
+VARIABLES
+  count,       \* The counter's current count
+  announced,   \* If the announcement has been made
+
+  signalled,   \* How often other prisoners have signalled
+
+  light,       \* Current status of the light
+
+  has_visited  \* The warden's view on who has actually been to the cell
+
+vars == <<count, announced, signalled, light, has_visited>>
+
+------------------------------------------------------------------------------
+
+\* The strategy differs slightly if the initial state of the light is
+\* known or not. If it's "off" then a simple count to N is
+\* sufficient. If the state is unknown we count to 2N - 1, and each
+\* prisoner will signal up to two times.
+
+SignalLimit == IF Light_Unknown THEN 2 ELSE 1
+
+VictoryThreshold ==
+  IF Light_Unknown
+  THEN Cardinality(Prisoner) * 2 - 1
+  ELSE Cardinality(Prisoner)
+
+------------------------------------------------------------------------------
+
+NormalPrisoner == Prisoner \ {Designated_Counter}
+
+\* The only thing to note here is that the count starts at one. Since
+\* only the counter will make an announcement if they are in the cell,
+\* this 1 means they have already counted themselves. We could also
+\* model this with a separate variable to note when the counter has
+\* visited the cell for the very first time, but this is not
+\* necessary.
+
+Init ==
+  /\ count = 1
+  /\ announced = FALSE
+  /\ signalled = [ p \in NormalPrisoner |-> 0 ]
+  /\ \/ Light_Unknown /\ light \in { "off", "on" }
+     \/ ~Light_Unknown /\ light = "off"
+  /\ has_visited = {}
+
+\* The action taken by the designated counter, if they are placed in
+\* the cell
+\*
+\* Note: we could simplify this change the IF to just be a
+\* precondition to the action, and while that would be a more elegant
+\* spec, I think this better models the decision procedure of the
+\* prisoner.
+CounterAction(p) ==
+  /\ p = Designated_Counter
+  /\ IF light = "on"
+     THEN
+       /\ light' = "off"
+       /\ count' = count + 1
+     ELSE
+       UNCHANGED <<light, count>>
+  /\ announced' = (count' >= VictoryThreshold)
+  /\ UNCHANGED <<signalled>>
+
+\* The action taken by the other prisoners, if they are placed in the
+\* cell
+\*
+\* Same note on the IF applies here.
+StandardAction(p) ==
+  /\ p \in NormalPrisoner
+  /\ IF light = "off" /\ signalled[p] < SignalLimit
+     THEN
+       /\ light' = "on"
+       /\ signalled' = [signalled EXCEPT ![p] = @ + 1]
+     ELSE
+       UNCHANGED <<light, signalled>>
+  /\ UNCHANGED <<count, announced>>
+
+\* The action performed by the warden: place a prisoner in the cell
+\* and maintain our own view of who's been selected (so we can judge
+\* if a victory announcement is correct)
+WardenAction(p) ==
```

Comment:
One technical reason for inlining `WardenAction` into `StandardAction` and `CounterAction` is that it allows TLC to produce more descriptive counterexamples, explicitly naming `StandardAction` and `CounterAction`. However, specifications should not be written solely to accommodate TLC.